sklearn/neighbors/_binary_tree.pxi

```diff
@@ -1334,14 +1335,16 @@ cdef class BinaryTree:
                 distances_npy = np.zeros(Xarr.shape[0], dtype='object')
                 for i in range(Xarr.shape[0]):
                     # make a new numpy array that wraps the existing data
-                    indices_npy[i] = cnp.PyArray_SimpleNewFromData(1, &counts[i], cnp.NPY_INTP, indices[i])
+                    # TODO: remove the explicit cast to cnp.intp_t* when cython min version >= 3.0
+                    indices_npy[i] = cnp.PyArray_SimpleNewFromData(1, <cnp.intp_t*>&counts[i], cnp.NPY_INTP, indices[i])
```

Comment:
This is another occurence of https://github.com/scikit-learn/scikit-learn/issues/25572#issuecomment-1451690983
This is fixed in cython 3.0. For now we can cast explicitely